File: examples3d/stress_tests/ragdolls3.rs
use rapier_testbed3d::TestbedViewer;
use rapier3d::prelude::*;

/// A 10-body articulated ragdoll (torso, head, 2×2 arm links, 2×2 leg links)
/// whose torso center is at `origin`. 9 joints, all with limits: spherical
/// neck, shoulders and hips; revolute elbows and knees.
fn ragdoll(world: &mut PhysicsWorld, origin: Vec3) {
    let part = |world: &mut PhysicsWorld, offset: Vec3, collider: ColliderBuilder| {
        let body = RigidBodyBuilder::dynamic().translation(origin + offset);
        let (handle, _) = world.insert(body, collider);
        (handle, offset)
    };

    let spherical = |parent: (RigidBodyHandle, Vec3),
                     child: (RigidBodyHandle, Vec3),
                     anchor: Vec3,
                     limit: f32| {
        SphericalJointBuilder::new()
            .local_anchor1(anchor - parent.1)
            .local_anchor2(anchor - child.1)
            .limits(JointAxis::AngX, [-limit, limit])
            .limits(JointAxis::AngY, [-limit, limit])
            .limits(JointAxis::AngZ, [-limit, limit])
            .contacts_enabled(false)
    };
    let revolute = |parent: (RigidBodyHandle, Vec3),
                    child: (RigidBodyHandle, Vec3),
                    anchor: Vec3,
                    limits: [f32; 2]| {
        RevoluteJointBuilder::new(Vec3::Z)
            .local_anchor1(anchor - parent.1)
            .local_anchor2(anchor - child.1)
            .limits(limits)
            .contacts_enabled(false)
    };

    let torso = part(world, Vec3::ZERO, ColliderBuilder::capsule_y(0.3, 0.15));
    let head = part(
        world,
        Vec3::new(0.0, 0.55, 0.0),
        ColliderBuilder::ball(0.15),
    );
    let neck = spherical(torso, head, Vec3::new(0.0, 0.42, 0.0), 0.5);
    world.insert_impulse_joint(torso.0, head.0, neck);

    for side in [-1.0f32, 1.0] {
        let upper_arm = part(
            world,
            Vec3::new(side * 0.36, 0.25, 0.0),
            ColliderBuilder::capsule_x(0.14, 0.06),
        );
        let forearm = part(
            world,
            Vec3::new(side * 0.70, 0.25, 0.0),
            ColliderBuilder::capsule_x(0.14, 0.06),
        );
        let thigh = part(
            world,
            Vec3::new(side * 0.09, -0.52, 0.0),
            ColliderBuilder::capsule_y(0.16, 0.07),
        );
        let shin = part(
            world,
            Vec3::new(side * 0.09, -0.92, 0.0),
            ColliderBuilder::capsule_y(0.16, 0.07),
        );

        let shoulder = spherical(torso, upper_arm, Vec3::new(side * 0.19, 0.25, 0.0), 1.2);
        world.insert_impulse_joint(torso.0, upper_arm.0, shoulder);
        let elbow = revolute(
            upper_arm,
            forearm,
            Vec3::new(side * 0.53, 0.25, 0.0),
            [0.0, 2.5],
        );
        world.insert_impulse_joint(upper_arm.0, forearm.0, elbow);
        let hip = spherical(torso, thigh, Vec3::new(side * 0.09, -0.33, 0.0), 1.0);
        world.insert_impulse_joint(torso.0, thigh.0, hip);
        let knee = revolute(thigh, shin, Vec3::new(side * 0.09, -0.72, 0.0), [0.0, 2.3]);
        world.insert_impulse_joint(thigh.0, shin.0, knee);
    }
}

pub async fn run(viewer: &mut TestbedViewer) -> anyhow::Result<()> {
    /*
     * World
     */
    let mut world = PhysicsWorld::new();

    /*
     * Ground
     */
    let (ground, _) = world.insert(
        RigidBodyBuilder::fixed().translation(Vec3::new(0.0, -1.0, 0.0)),
        ColliderBuilder::cuboid(100.0, 1.0, 100.0),
    );
    let _ = ground;

    /*
     * Ragdolls dropped into a pile: a 5x5 grid, 5 layers (125 ragdolls,
     * 1250 dynamic bodies, 1125 limit joints).
     */
    for layer in 0..5 {
        for row in 0..5 {
            for col in 0..5 {
                let origin =
                    Vec3::new(col as f32 * 2.2, 1.5 + layer as f32 * 2.6, row as f32 * 2.2);
                ragdoll(&mut world, origin);
            }
        }
    }

    /*
     * Set up the testbed.
     */
    viewer.set_world(&mut world);
    viewer.look_at(Vec3::new(-12.0, 10.0, -12.0), Vec3::new(4.5, 1.0, 4.5));

    while viewer.render_frame(&mut world).await {
        if viewer.simulating() {
            world.step();
        }
    }
    Ok(())
}

File: c/testbed/examples3d/stress_tests/ragdolls3.c
/* Port of examples3d/stress_tests/ragdolls3.rs. */
#include "testbed.h"
#include "rapier_helpers.h"
#include "rapier_math.h"

typedef struct Part {
    R3RigidBodyHandle handle;
    R3Vector offset;
} Part;

/* A body and its offset from the ragdoll's torso. */
static Part part(R3World *world, R3Vector origin, R3Vector offset, R3ColliderDesc collider,
                 int noSleep) {
    R3RigidBodyDesc body = r3DynamicRigidBodyDesc();
    body.position.translation = r3VectorAdd(origin, offset);
    body.canSleep = !noSleep;
    Part result = {.offset = offset};
    result.handle = r3InsertRigidBody(world, &body);
    r3InsertCollider(result.handle, &collider);

    return result;
}

static R3JointDesc revolute(Part parent, Part child, R3Vector anchor, R3Real min, R3Real max) {
    R3JointDesc joint = r3RevoluteJointDesc(r3Vector(0, 0, 1));
    joint.localFrame1.translation = r3VectorSub(anchor, parent.offset);
    joint.localFrame2.translation = r3VectorSub(anchor, child.offset);
    r3JointDesc_SetLimits(&joint, R3_AXIS_ANG_X, min, max);
    joint.contactsEnabled = 0;
    return joint;
}

static R3JointDesc spherical(Part parent, Part child, R3Vector anchor, R3Real limit) {
    R3JointDesc joint = r3SphericalJointDesc();
    joint.localFrame1.translation = r3VectorSub(anchor, parent.offset);
    joint.localFrame2.translation = r3VectorSub(anchor, child.offset);
    r3JointDesc_SetLimits(&joint, R3_AXIS_ANG_X, -limit, limit);
    r3JointDesc_SetLimits(&joint, R3_AXIS_ANG_Y, -limit, limit);
    r3JointDesc_SetLimits(&joint, R3_AXIS_ANG_Z, -limit, limit);
    joint.contactsEnabled = 0;
    return joint;
}

/* Ten bodies and nine limited joints, matching the Rust ragdoll. */
static void ragdoll(R3World *world, R3Vector origin, int noSleep) {
    R3ColliderDesc collider = r3CapsuleYColliderDesc(.3, .15);
    Part torso = part(world, origin, r3Vector(0, 0, 0), collider, noSleep);
    collider = r3BallColliderDesc(.15);
    Part head = part(world, origin, r3Vector(0, 0.55, 0), collider, noSleep);
    R3JointDesc neck = spherical(torso, head, r3Vector(0, 0.42, 0), .5);
    r3InsertImpulseJoint(torso.handle, head.handle, &neck);

    for (int side = -1; side <= 1; side += 2) {
        collider = r3CapsuleXColliderDesc(.14, .06);
        Part upperArm = part(world, origin, r3Vector(side * .36, .25, 0), collider, noSleep);
        collider = r3CapsuleXColliderDesc(.14, .06);
        Part forearm = part(world, origin, r3Vector(side * .70, .25, 0), collider, noSleep);
        collider = r3CapsuleYColliderDesc(.16, .07);
        Part thigh = part(world, origin, r3Vector(side * .09, -.52, 0), collider, noSleep);
        collider = r3CapsuleYColliderDesc(.16, .07);
        Part shin = part(world, origin, r3Vector(side * .09, -.92, 0), collider, noSleep);
        R3JointDesc shoulder = spherical(torso, upperArm, r3Vector(side * .19, .25, 0), 1.2);
        r3InsertImpulseJoint(torso.handle, upperArm.handle, &shoulder);

        R3JointDesc elbow = revolute(upperArm, forearm, r3Vector(side * .53, .25, 0), 0, 2.5);
        r3InsertImpulseJoint(upperArm.handle, forearm.handle, &elbow);

        R3JointDesc hip = spherical(torso, thigh, r3Vector(side * .09, -.33, 0), 1);
        r3InsertImpulseJoint(torso.handle, thigh.handle, &hip);

        R3JointDesc knee = revolute(thigh, shin, r3Vector(side * .09, -.72, 0), 0, 2.3);
        r3InsertImpulseJoint(thigh.handle, shin.handle, &knee);
    }
}

void tbStressTestsRagdolls3(Testbed *testbed) {
    R3World *world = r3NewWorld();
    {
        R3RigidBodyDesc rigidBody = r3FixedRigidBodyDesc();
        rigidBody.position.translation = r3Vector(0, -1, 0);
        rigidBody.canSleep = !testbed->noSleep;
        R3ColliderDesc collider = r3CuboidColliderDesc(r3Vector(100, 1, 100));
        R3RigidBodyHandle rigidBodyHandle = r3InsertRigidBody(world, &rigidBody);
        r3InsertCollider(rigidBodyHandle, &collider);
    }
    /* 125 ragdolls: a 5 by 5 grid with 5 layers. */
    for (int layer = 0; layer < 5; ++layer) {
        for (int row = 0; row < 5; ++row) {
            for (int col = 0; col < 5; ++col) {
                ragdoll(world, r3Vector(col * 2.2, 1.5 + layer * 2.6, row * 2.2), testbed->noSleep);
            }
        }
    }
    tbCamera(testbed, -12, 10, -12, 4.5, 1, 4.5);

    tbSetWorld(testbed, world);

    while (tbRenderFrame(testbed, &world)) {
        if (tbSimulating(testbed)) {
            r3Step(world, NULL, NULL);
        }
    }
    r3FreeWorld(world);
}
